icu4c/source/common/uloc_keytype.cpp

```diff
@@ -229,13 +214,23 @@ initFromResourceBundle(UErrorCode& sts) {
                     // e.g. America:Los_Angeles
                     if (uprv_strchr(legacyTypeId, ':') != NULL) {
                         int32_t legacyTypeIdLen = static_cast<int32_t>(uprv_strlen(legacyTypeId));
-                        char* legacyTypeIdBuf = (char*)uprv_malloc(legacyTypeIdLen + 1);
+                        icu::CharString* legacyTypeIdBuf =
+                                gKeyTypeStringPool->create();
                         if (legacyTypeIdBuf == NULL) {
                             sts = U_MEMORY_ALLOCATION_ERROR;
                             break;
                         }
+                        int32_t resultCapacity;
+                        char* buffer = legacyTypeIdBuf->getAppendBuffer(
```

Comment:
Differently clunky... I suggest you allocate a CharString that copies the legacyTypeId, then modify its data() buffer replacing any ':' with '/'. data() returns a non-const pointer if "this" is not const. No need for getAppendBuffer(), no need to copy each character separately.

Reply:
Done.